### ISSUE TITLE
Repeated preprocessor in pipeline: ZeroCount()

### ISSUE THREAD

**Issue by dah33** [EpistasisLab/tpot#564]
I've been experimenting with TPOT on [Fashion MNIST](https://github.com/zalandoresearch/fashion-mnist), and been presented with this rather strange pipeline:

```
exported_pipeline = make_pipeline(
    ZeroCount(),
    ZeroCount(),
    ZeroCount(),
    ZeroCount(),
    LogisticRegression(C=0.0001, dual=True)
)
```

ZeroCount just counts the number of zero and non-zero features per sample. Since there are some zero and some non-zero features in every sample, repeating ZeroCount will just report the same zero count each time, and a non-zero count that increases by two each time. I can't see how this would help an ML algorithm.

Would it be possible for the grammar to collapse repeated ZeroCount steps?

Here's my code;

```
from tpot import TPOTClassifier
from sklearn.model_selection import train_test_split
import pandas as pd

def load_mnist(path, kind='train'):
    import os
    import gzip
    import numpy as np

    """Load MNIST data from `path`"""
    labels_path = os.path.join(path,
                               '%s-labels-idx1-ubyte.gz'
                               % kind)
    images_path = os.path.join(path,
                               '%s-images-idx3-ubyte.gz'
                               % kind)

    with gzip.open(labels_path, 'rb') as lbpath:
        labels = np.frombuffer(lbpath.read(), dtype=np.uint8,
                               offset=8)

    with gzip.open(images_path, 'rb') as imgpath:
        images = np.frombuffer(imgpath.read(), dtype=np.uint8,
                               offset=16).reshape(len(labels), 784)

    return images, labels

#X_train, y_train = load_mnist('fashion-mnist/data/fashion', kind='train')
#X_test, y_test = load_mnist('fashion-mnist/data/fashion', kind='t10k')
data, target = load_mnist('fashion-mnist/data/fashion', kind='t10k')


X_train, X_test, y_train, y_test = train_test_split(data, target,
                                                    train_size=0.15,test_size=0.05)

tpot = TPOTClassifier(generations=100, population_size=10, verbosity=2,
                      config_dict='TPOT light')

tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
tpot.export('tpot_fashion_pipeline_light.py')
```

**Comment by weixuanfu** [comment]
Thank you for reporting this issue. I think we need add a rule in TPOT grammar to collapse repeated preprocessor operators in a pipeline. @rhiever

**Comment by rhiever** [comment]
@dah33, have you tried sequentially moving the ZeroCount operators to see how it affects the prediction accuracy? The optimization procedure has a pressure to maximize prediction accuracy while minimizing the number of operators in the pipeline, so the ZeroCount operators should only be there if they improved the 10-fold CV score.

How many generations and what population size did you use for this run?

**Comment by dah33** [comment]
Hi. I did try removing the ZeroCount sequentially. There test metrics are a bit noisy, but I do not see any discernable improvement in performance with each additional ZeroCount(). I think it may be getting lucky with the noise at each step, tricking it into adding ZeroCount preprocessors.

**Comment by rhiever** [comment]
That's possible. It's also possible that TPOT didn't have enough iterations to prune down the pipeline to get rid of the superfluous ZeroCount operators. For example, if TPOT was run only for 5 generations or so.

**Comment by dah33** [comment]
My code is above. 100 generations, 10 population size, but not a huge training set. 1500 observations for training, 500 for validation.

**Comment by rhiever** [comment]
Thanks! Increasing the population size will be pretty important for TPOT performance; if I had to choose, I would take 100 population and 10 generations instead of 10 pop/100 gen. Of course, 100/100 would be even better.

**Comment by dah33** [comment]
Thank you.
